File: bot/src/music/ytlibrary.ts
import type Database from "better-sqlite3";
import path from "node:path";
import { existsSync, readdirSync } from "node:fs";
import type { Logger } from "../logger.js";

export interface SongMeta {
  name: string;
  artist: string;
  duration?: number;
}

/** Local provider surface used by YtLibrary (save dir + post-save re-index). */
export interface YtLibraryLocalSource {
  getMusicDir(): string;
  refresh(): Promise<unknown>;
}

/** YouTube provider surface used by YtLibrary (background MP3 download). */
export interface YtLibraryYoutubeSource {
  downloadAudioMp3(videoId: string, outDir: string, baseName: string): Promise<string>;
}

export interface YtLibraryOptions {
  db: Database.Database;
  /** Absolute music library root (saved MP3s go in a subdir so LocalProvider indexes them). */
  musicDir: string;
  /** Download a video's audio as MP3 into `<outDir>/<baseName>.mp3`; returns the path. */
  download: (videoId: string, outDir: string, baseName: string) => Promise<string>;
  /** Re-index the local library after a save so the new track becomes searchable. */
  refresh: () => Promise<unknown>;
  logger?: Logger;
}

/** Subdir under MUSIC_DIR holding saved YouTube tracks (still part of the indexed library). */
const SAVE_SUBDIR = "youtube";

/**
 * YouTube → permanent local library (ROADMAP "Adjacent feature"). Maps a
 * canonical video id → a saved, tagged MP3 in the library, so replaying the same
 * video (any URL form) serves the local file instead of re-streaming/-downloading.
 * Saves run in the background (stream-first UX) behind an in-flight lock; a failed
 * save never affects playback. Backed by the bot's existing SQLite db.
 */
export class YtLibrary {
  private db: Database.Database;
  private musicDir: string;
  private download: YtLibraryOptions["download"];
  private refresh: YtLibraryOptions["refresh"];
  private logger?: Logger;
  private inFlight = new Set<string>();
  private getStmt: Database.Statement;
  private addStmt: Database.Statement;

  constructor(opts: YtLibraryOptions) {
    this.db = opts.db;
    this.musicDir = opts.musicDir;
    this.download = opts.download;
    this.refresh = opts.refresh;
    this.logger = opts.logger;

    this.db.exec(`
      CREATE TABLE IF NOT EXISTS yt_saved (
        video_id TEXT PRIMARY KEY,
        path TEXT NOT NULL,
        title TEXT NOT NULL,
        artist TEXT NOT NULL,
        duration INTEGER,
        created_at INTEGER NOT NULL
      );
    `);
    this.getStmt = this.db.prepare(`SELECT path FROM yt_saved WHERE video_id = ?`);
    this.addStmt = this.db.prepare(
      `INSERT OR REPLACE INTO yt_saved (video_id, path, title, artist, duration, created_at) VALUES (?, ?, ?, ?, ?, ?)`,
    );
  }

  /** Saved local MP3 path for a video id, if saved AND the file still exists; else null. */
  lookup(videoId: string): string | null {
    if (!videoId) return null;
    const row = this.getStmt.get(videoId) as { path: string } | undefined;
    if (row?.path && existsSync(row.path)) return row.path;

    const onDisk = findSavedOnDisk(this.musicDir, videoId);
    if (onDisk) {
      this.reconcileDiskHit(videoId, onDisk);
      return onDisk;
    }
    return null;
  }

  /** Register an on-disk save the DB missed (e.g. after manual copy or DB reset). */
  private reconcileDiskHit(videoId: string, filePath: string): void {
    const row = this.getStmt.get(videoId) as { path: string } | undefined;
    if (row?.path && existsSync(row.path)) return;
    const meta = parseSavedFilename(path.basename(filePath), videoId);
    this.addStmt.run(videoId, filePath, meta.title, meta.artist, 0, Date.now());
  }

  /**
   * Download + save a video as a tagged MP3 in the background (fire-and-forget).
   * No-op if already saved or a save is in-flight. Errors are swallowed (logged) —
   * a failed save must never affect the already-playing stream.
   */
  saveInBackground(videoId: string, meta: SongMeta): void {
    if (!videoId || this.lookup(videoId) || this.inFlight.has(videoId)) return;
    this.inFlight.add(videoId);
    const outDir = path.join(this.musicDir, SAVE_SUBDIR);
    const base = sanitizeBase(meta, videoId);
    void (async () => {
      try {
        this.logger?.info({ videoId, title: meta.name }, "YT save: downloading");
        const savedPath = await this.download(videoId, outDir, base);
        this.addStmt.run(videoId, savedPath, meta.name || videoId, meta.artist || "YouTube", meta.duration ?? 0, Date.now());
        await this.refresh();
        this.logger?.info({ videoId, path: savedPath }, "YT save: saved to library");
      } catch (err) {
        this.logger?.warn({ err, videoId }, "YT save failed (kept streaming)");
      } finally {
        this.inFlight.delete(videoId);
      }
    })();
  }
}

/** Find a saved YouTube MP3 on disk by the `[videoId]` tag in its filename. */
export function findSavedOnDisk(musicDir: string, videoId: string): string | null {
  if (!videoId) return null;
  const dir = path.join(musicDir, SAVE_SUBDIR);
  if (!existsSync(dir)) return null;

  const tag = `[${videoId}]`;
  try {
    for (const name of readdirSync(dir)) {
      if (!name.toLowerCase().endsWith(".mp3")) continue;
      if (!name.includes(tag)) continue;
      const full = path.join(dir, name);
      if (existsSync(full)) return full;
    }
  } catch {
    return null;
  }
  return null;
}

/** Parse `<artist> - <title> [videoId].mp3` written by {@link sanitizeBase}. */
export function parseSavedFilename(basename: string, videoId: string): { title: string; artist: string } {
  const escaped = videoId.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const m = basename.match(new RegExp(`^(.+?) - (.+) \\[${escaped}\\]\\.mp3$`, "i"));
  if (m) {
    return { artist: m[1].trim(), title: m[2].trim() };
  }
  return { title: videoId, artist: "YouTube" };
}

/** Build a filesystem-safe base name: `<artist> - <title> [videoId]`. */
export function sanitizeBase(meta: SongMeta, videoId: string): string {
  const raw = [meta.artist, meta.name].filter(Boolean).join(" - ") || videoId;
  const safe = raw.replace(/[\/\\?%*:|"<>]/g, "").replace(/\s+/g, " ").trim().slice(0, 150);
  return `${safe || videoId} [${videoId}]`;
}

File: crates/mp-music/src/ytlibrary.rs
// Copyright (c) 2026 Lane Ambrose
// SPDX-License-Identifier: MIT

//! YouTube → local library MP3. Failed saves never affect the playing stream.

use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};
use mp_db::Database;

use crate::blacklist::extract_video_id;
use crate::youtube::YoutubeClient;

pub const SAVE_SUBDIR: &str = "youtube";

pub struct YtLibrary {
    db: Arc<Database>,
    music_dir: PathBuf,
    youtube: YoutubeClient,
    in_flight: Arc<Mutex<std::collections::HashSet<String>>>,
}

impl YtLibrary {
    pub fn new(db: Arc<Database>, music_dir: impl AsRef<Path>, youtube: YoutubeClient) -> Self {
        Self {
            db,
            music_dir: music_dir.as_ref().to_path_buf(),
            youtube,
            in_flight: Arc::new(Mutex::new(std::collections::HashSet::new())),
        }
    }

    pub fn lookup(&self, video_id: &str) -> Option<String> {
        let video_id = extract_video_id(video_id)?;
        if let Ok(Some(p)) = self.db.yt_saved().lookup(&video_id) {
            if Path::new(&p).is_file() {
                return Some(p);
            }
        }
        find_saved_on_disk(&self.music_dir, &video_id)
    }

    pub fn save_in_background(&self, video_id: String, title: String, artist: String, duration: u32) {
        let Some(video_id) = extract_video_id(&video_id) else {
            // X/Bandcamp page URLs are not safe -o template stems.
            return;
        };
        if self.lookup(&video_id).is_some() {
            return;
        }
        {
            let mut g = self.in_flight.lock().expect("yt inflight");
            if !g.insert(video_id.clone()) {
                return;
            }
        }
        let db = Arc::clone(&self.db);
        let music_dir = self.music_dir.clone();
        let youtube = self.youtube.clone();
        let inflight = Arc::clone(&self.in_flight);
        let vid = video_id.clone();
        std::thread::spawn(move || {
            let out_dir = music_dir.join(SAVE_SUBDIR);
            let base = sanitize_base(&artist, &title, &vid);
            match youtube.download_audio_mp3(&vid, &out_dir, &base) {
                Some(path) => {
                    if let Err(e) = db.yt_saved().insert(&vid, &path, &title, &artist, duration as i64)
                    {
                        tracing::warn!(error = %e, video_id = %vid, "YT save: db insert failed");
                    } else {
                        tracing::info!(video_id = %vid, path = %path, "YT save: saved to library");
                    }
                }
                None => tracing::warn!(video_id = %vid, "YT save failed (kept streaming)"),
            }
            inflight.lock().expect("yt inflight").remove(&vid);
        });
    }
}

pub fn find_saved_on_disk(music_dir: &Path, video_id: &str) -> Option<String> {
    let video_id = extract_video_id(video_id)?;
    let dir = music_dir.join(SAVE_SUBDIR);
    let tag = format!("[{video_id}]");
    let rd = std::fs::read_dir(&dir).ok()?;
    for ent in rd.flatten() {
        let name = ent.file_name();
        let name = name.to_string_lossy();
        if !name.to_ascii_lowercase().ends_with(".mp3") {
            continue;
        }
        if name.contains(&tag) {
            let full = ent.path();
            if full.is_file() {
                return Some(full.to_string_lossy().into_owned());
            }
        }
    }
    None
}

pub fn sanitize_base(artist: &str, title: &str, video_id: &str) -> String {
    let raw = [artist, title]
        .iter()
        .filter(|s| !s.is_empty())
        .copied()
        .collect::<Vec<_>>()
        .join(" - ");
    let raw = if raw.is_empty() { video_id } else { &raw };
    let safe: String = raw
        .chars()
        .map(|c| match c {
            '/' | '\\' | '?' | '%' | '*' | ':' | '|' | '"' | '<' | '>' => ' ',
            other => other,
        })
        .collect::<String>()
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ");
    let safe = if safe.len() > 150 {
        safe.chars().take(150).collect()
    } else {
        safe
    };
    let id = extract_video_id(video_id).unwrap_or_else(|| {
        video_id
            .chars()
            .filter(|c| c.is_ascii_alphanumeric() || *c == '-' || *c == '_')
            .take(11)
            .collect()
    });
    if id.is_empty() {
        return safe;
    }
    format!("{} [{id}]", if safe.is_empty() { id.as_str() } else { &safe })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sanitizes_filename() {
        let b = sanitize_base("A/B", "Hi?", "hLOheGDwD_0");
        assert!(b.contains("[hLOheGDwD_0]"));
        assert!(!b.contains('/'));
        assert!(!b.contains('?'));
    }

    #[test]
    fn save_id_rejects_path_injection() {
        assert!(extract_video_id("https://x.com/../../../../tmp/pwn").is_none());
        assert!(extract_video_id("https://x.com/%(id)s").is_none());
        let b = sanitize_base("a", "b", "https://x.com/../../tmp/x");
        assert!(!b.contains(".."));
        assert!(!b.contains('/'));
        assert!(!b.contains('%'));
    }
}
